app/services/fdc_backorderer.rb

```diff
@@ -0,0 +1,133 @@
+# frozen_string_literal: true
+
+# Place and update orders based on missing stock.
+class FdcBackorderer
+  attr_reader :user, :urls
+
+  def initialize(user, urls)
+    @user = user
+    @urls = urls
+  end
+
+  def find_or_build_order(ofn_order)
+    find_open_order || build_new_order(ofn_order)
+  end
+
+  def build_new_order(ofn_order)
+    OrderBuilder.new_order(ofn_order, urls.orders_url).tap do |order|
+      order.saleSession = build_sale_session(ofn_order)
+    end
+  end
+
+  def find_open_order
+    graph = import(urls.orders_url)
+    open_orders = graph&.select do |o|
+      o.semanticType == "dfc-b:Order" && o.orderStatus[:path] == "Held"
+    end
+
+    return if open_orders.blank?
+
+    # If there are multiple open orders, we don't know which one to choose.
+    # We want the order we placed for the same distributor in the same order
```

Comment:
Ok. I'm pretty sure the Shopify Orders endpoint will display old (completed orders), we should only have 1 __Draft__ order though.

The Shopify Hub app is storing the ID of the Order returned from the initial POST request, to utilise in future PUT requests. Could we do that ?

Reply:
We could. And maybe we need to. I tried to get around it to have minimal local data. Local data structures increase maintenance and can become out of sync. And I was worried about longer dev times as well. But I agree that the current solution isn't great.